File: 4258-construct-uniform-parity-array-ii/construct-uniform-parity-array-ii.cpp
#include <vector>
#include <algorithm>
#include <climits>

using namespace std;

class Solution {
public:
    bool uniformArray(vector<int>& nums1) {
        int minOdd = INT_MAX;
        int minEven = INT_MAX;

        for (int x : nums1) {
            if (x % 2 != 0) {
                minOdd = min(minOdd, x);
            } else {
                minEven = min(minEven, x);
            }
        }

        // If array already contains only odds or only evens
        if (minOdd == INT_MAX || minEven == INT_MAX) {
            return true;
        }

        // All elements can only become odd if every even number is strictly greater than minOdd
        return minOdd < minEven;
    }
};

File: 4258-construct-uniform-parity-array-ii/construct-uniform-parity-array-ii.java
class Solution {
    public boolean uniformArray(int[] nums1) {
        int minOdd = Integer.MAX_VALUE;
        int minEven = Integer.MAX_VALUE;

        for (int x : nums1) {
            if (x % 2 != 0) {
                minOdd = Math.min(minOdd, x);
            } else {
                minEven = Math.min(minEven, x);
            }
        }

        // If all numbers are already of the same parity, it's always possible
        if (minOdd == Integer.MAX_VALUE || minEven == Integer.MAX_VALUE) {
            return true;
        }

        // To make all elements odd: every even number x must satisfy x - minOdd >= 1 (x > minOdd)
        // If minEven < minOdd, we cannot make all elements odd or even
        return minOdd < minEven;
    }
}